include/lsst/partition/ObjectIndex.h

```diff
@@ -0,0 +1,195 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+/// \file
+/// \brief Operations with the "secondary" index
+
+#ifndef LSST_PARTITION_OBJECTINDEX_H
+#define LSST_PARTITION_OBJECTINDEX_H
+
+// System headers
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+
+// LSST headers
+#include "lsst/partition/Csv.h"
+
+// Forward declarations
+namespace lsst {
+namespace partition {
+    class ChunkLocation;
+}} // namespace lsst::partition
+
+// This header declarations
+namespace lsst {
+namespace partition {
+
+/**
+ * Class ObjectIndex is a singleton serving as a front-end to a specific implementation of
+ * the "secondary" index.
+ * 
+ * Objects of the class can be open in two modes:
+ * - WRITE: for writing into an index file at the specified location.
+ * - READ: reading from an index via the specified source.
+ * 
+ * Files written in the WRITE mode are open in the "append" mode, hence their previoous
+ * content won't be truncated. It's up to a client to ensure the files get removed if
+ * the clean starting state is required.
+ * 
+ * The source specification in the READ node is a URL-style resource, which has the following
+ * general syntax:
+ * @code
+ *   <technology>:<technolog-specific-details>
+ * @code
+ * Presently, only the file-based technology is supported. Hence, the file URL is expected
+ * to look like this:
+ * @code
+ *   file:<path-to-the-index-file>
+ * @code
+ * Normally the path specified in the resource would be a file which was produced
+ * in the WRITE mode. The path could have a relative or absolute value. It could also be
+ * located on a local or remote filesystem. Since the file would be read sequentially then
+ * no specific requirements for file locking or direct access are imposed by this implemenation.
+ *
+ * Notes on the thread-safety of the class's implementation:
+ * 1) All public methods of the class are thread safe.
+ * 2) An internal state of the index is guarded by a mutex.
+ */
+class ObjectIndex {
+public:
+    // Objects of this class can't be copied or directly constructed by clients.
+    ObjectIndex(ObjectIndex const&) = delete;
+    ObjectIndex& operator=(ObjectIndex const&) = delete;
+
+    /// The non-trivial destructor is needed to close and release resources used by the index.
+    ~ObjectIndex();
+
+    /// Modes for opening the index.
+    enum Mode {
+        READ,
+        WRITE
+    };
+
+    /**
+     * The factory method.
+     * 
+     * @return  A reference to the underlining implementation of the index
+     */
+    static ObjectIndex& instance() {
+        static ObjectIndex oi;
```

Comment:
With your use pattern for this singleton I'd actually recommend to never have a singleton instance in non-open/unusable state, one of the two options could work for that:

- get rid of singleton completely (better) 
- replace static instance with a global pointer, open()/create() methods would initialize it, instance() would throw if pointer is nullptr.